File: app/services/openai_service.py
from openai import OpenAI
from google import genai
from google.genai import types
from groq import Groq
from app.config import get_settings
from app.models.schemas import (
    TextPostRequest, TextPostResponse,
    ImagePostRequest, ImagePostResponse,
    VideoScriptRequest, VideoScriptResponse,
    ContentStats
)
import json
import asyncio


class OpenAIService:
    def __init__(self):
        settings = get_settings()
        self.openai_client = None
        self.gemini_client = None
        self.groq_client = None

        # Try OpenAI first
        if settings.openai_api_key:
            try:
                self.openai_client = OpenAI(api_key=settings.openai_api_key)
                print("OpenAI client initialized")
            except Exception as e:
                print(f"OpenAI init failed: {e}")

        # Setup Gemini
        if settings.google_api_key:
            try:
                self.gemini_client = genai.Client(
                    api_key=settings.google_api_key)
                print("Gemini client initialized")
            except Exception as e:
                print(f"Gemini init failed: {e}")

        # Setup Groq as final fallback
        if settings.groq_api_key:
            try:
                self.groq_client = Groq(api_key=settings.groq_api_key)
                print("Groq client initialized")
            except Exception as e:
                print(f"Groq init failed: {e}")

    def _get_platform_context(self, platform: str) -> dict:
        """Get platform-specific limits and best practices."""
        platforms = {
            "instagram": {
                "char_limit": 2200,
                "hashtag_limit": 30,
                "best_practices": "Use emojis, line breaks, and storytelling. End with a CTA."
            },
            "linkedin": {
                "char_limit": 3000,
                "hashtag_limit": 5,
                "best_practices": "Professional tone, use bullet points, share insights and value."
            },
            "twitter": {
                "char_limit": 280,
                "hashtag_limit": 3,
                "best_practices": "Be concise, use hooks, create threads for longer content."
            },
            "facebook": {
                "char_limit": 63206,
                "hashtag_limit": 10,
                "best_practices": "Tell stories, ask questions, encourage engagement."
            },
            "tiktok": {
                "char_limit": 2200,
                "hashtag_limit": 10,
                "best_practices": "Trendy, casual, use popular hashtags and hooks."
            },
            "youtube": {
                "char_limit": 5000,
                "hashtag_limit": 15,
                "best_practices": "SEO-optimized descriptions, include timestamps and links."
            }
        }
        return platforms.get(platform, platforms["instagram"])

    async def _generate_with_openai(self, system_prompt: str, user_prompt: str) -> dict:
        """Generate content using OpenAI."""
        response = await asyncio.to_thread(
            self.openai_client.chat.completions.create,
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt}
            ],
            temperature=0.8,
            max_tokens=1000,
            response_format={"type": "json_object"}
        )
        return json.loads(response.choices[0].message.content)

    async def _generate_with_gemini(self, system_prompt: str, user_prompt: str) -> dict:
        """Generate content using Google Gemini SDK."""
        full_prompt = f"{system_prompt}\n\n{user_prompt}"

        response = await asyncio.to_thread(
            self.gemini_client.models.generate_content,
            model="gemini-2.0-flash",
            contents=full_prompt,
            config=types.GenerateContentConfig(
                temperature=0.8,
                max_output_tokens=1000,
                response_mime_type="application/json"
            )
        )

        text = response.text.strip()
        return self._clean_json_response(text)

    async def _generate_with_groq(self, system_prompt: str, user_prompt: str) -> dict:
        """Generate content using Groq (Llama 3.1)."""
        full_system = system_prompt + \
            "\n\nIMPORTANT: Respond ONLY with valid JSON. No markdown, no code blocks, no extra text."

        response = await asyncio.to_thread(
            self.groq_client.chat.completions.create,
            model="llama-3.1-8b-instant",
            messages=[
                {"role": "system", "content": full_system},
                {"role": "user", "content": user_prompt}
            ],
            temperature=0.8,
            max_tokens=1024,
            response_format={"type": "json_object"}
        )

        text = response.choices[0].message.content.strip()
        return self._clean_json_response(text)

    def _clean_json_response(self, text: str) -> dict:
        """Clean and parse JSON response."""
        # Remove markdown code blocks if present
        if text.startswith("```json"):
            text = text[7:]
        if text.startswith("```"):
            text = text[3:]
        if text.endswith("```"):
            text = text[:-3]
        return json.loads(text.strip())

    async def _generate_content(self, system_prompt: str, user_prompt: str) -> dict:
        """Try OpenAI → Gemini → Groq with fallback."""

        # Try OpenAI first
        if self.openai_client:
            try:
                print("Trying OpenAI...")
                return await self._generate_with_openai(system_prompt, user_prompt)
            except Exception as e:
                print(f"OpenAI failed: {e}")

        # Fallback to Gemini
        if self.gemini_client:
            for attempt in range(2):
                try:
                    print(f"Using Gemini (attempt {attempt + 1}/2)...")
                    if attempt > 0:
                        await asyncio.sleep(2)
                    return await self._generate_with_gemini(system_prompt, user_prompt)
                except Exception as e:
                    print(f"Gemini attempt {attempt + 1} failed: {e}")

        # Final fallback to Groq
        if self.groq_client:
            for attempt in range(2):
                try:
                    print(f"Using Groq (attempt {attempt + 1}/2)...")
                    if attempt > 0:
                        await asyncio.sleep(2)
                    return await self._generate_with_groq(system_prompt, user_prompt)
                except Exception as e:
                    print(f"Groq attempt {attempt + 1} failed: {e}")
                    if attempt == 1:
                        raise Exception(
                            f"All AI providers failed. Last error: {e}")

        raise Exception(
            "No AI provider available. Configure OpenAI, Google, or Groq API key.")

    async def generate_text_post(self, request: TextPostRequest) -> TextPostResponse:
        """Generate text post content."""

        platform_ctx = self._get_platform_context(request.platform.value)

        system_prompt = f"""You are an expert social media content creator.
Create engaging content for {request.platform.value}.

Platform Guidelines:
- Character limit: {platform_ctx['char_limit']}
- Hashtag limit: {platform_ctx['hashtag_limit']}
- Best practices: {platform_ctx['best_practices']}

Content Requirements:
- Tone: {request.tone.value}
- Goal: {request.goal.value}
- Include emojis: {request.includeEmojis}
- Generate hashtags: {request.includeHashtags}
- Include CTA: {request.includeCTA}

Respond with valid JSON:
{{
    "caption": "The main caption text",
    "hashtags": ["#hashtag1", "#hashtag2"],
    "cta": "Call to action text"
}}"""

        user_prompt = f"""Create a {request.tone.value} social media post for {request.platform.value} about:

{request.topic}

Goal: {request.goal.value}. Make it engaging and optimized for the platform."""

        result = await self._generate_content(system_prompt, user_prompt)

        caption = result.get("caption", "")
        words = len(caption.split())

        return TextPostResponse(
            success=True,
            caption=caption,
            hashtags=result.get(
                "hashtags", []) if request.includeHashtags else [],
            cta=result.get("cta", "") if request.includeCTA else "",
            stats=ContentStats(
                characters=len(caption),
                words=words,
                readTime=f"{max(1, words // 3)} sec",
                engagementScore="High" if len(
                    result.get("hashtags", [])) >= 5 else "Medium"
            ),
            creditsUsed=1,
            creditsRemaining=149
        )

    async def generate_image_post(self, request: ImagePostRequest) -> ImagePostResponse:
        """Generate image post with caption and image prompt."""

        system_prompt = f"""You are an expert social media content creator.
Create engaging content for {request.platform.value} with an image.

Respond with valid JSON:
{{
    "caption": "The caption text",
    "hashtags": ["#hashtag1", "#hashtag2"],
    "cta": "Call to action",
    "imagePrompt": "Detailed prompt for AI image generation"
}}"""

        user_prompt = f"""Create a {request.tone.value} image post for {request.platform.value} about:

{request.topic}

Goal: {request.goal.value}. Include a detailed image generation prompt."""

        result = await self._generate_content(system_prompt, user_prompt)

        caption = result.get("caption", "")
        words = len(caption.split())

        return ImagePostResponse(
            success=True,
            caption=caption,
            hashtags=result.get("hashtags", []),
            cta=result.get("cta", ""),
            imagePrompt=result.get("imagePrompt", ""),
            stats=ContentStats(
                characters=len(caption),
                words=words,
                readTime=f"{max(1, words // 3)} sec",
                engagementScore="High"
            ),
            creditsUsed=2,
            creditsRemaining=23
        )

    async def generate_video_script(self, request: VideoScriptRequest) -> VideoScriptResponse:
        """Generate video script."""

        system_prompt = f"""You are an expert video content creator.
Create a {request.duration} video script for {request.platform.value}.

Respond with valid JSON:
{{
    "hook": "Attention-grabbing opening line",
    "script": "Main video script content",
    "cta": "Call to action at the end",
    "hashtags": ["#hashtag1", "#hashtag2"]
}}"""

        user_prompt = f"""Create a {request.tone.value} video script for {request.platform.value} about:

{request.topic}

Duration: {request.duration}
Goal: {request.goal.value}"""

        result = await self._generate_content(system_prompt, user_prompt)

        script = result.get("script", "")
        words = len(script.split())

        return VideoScriptResponse(
            success=True,
            hook=result.get("hook", ""),
            script=script,
            cta=result.get("cta", ""),
            hashtags=result.get("hashtags", []),
            stats=ContentStats(
                characters=len(script),
                words=words,
                readTime=f"{max(1, words // 3)} sec",
                engagementScore="High"
            ),
            creditsUsed=3,
            creditsRemaining=7
        )


# Singleton instance
_openai_service = None


def get_openai_service() -> OpenAIService:
    global _openai_service
    if _openai_service is None:
        _openai_service = OpenAIService()
    return _openai_service

File: services/openaiService.js
import OpenAI from 'openai';
import { GoogleGenerativeAI } from '@google/generative-ai';
import Groq from 'groq-sdk';
import { getSettings } from '../config/index.js';

class OpenAIService {
    constructor() {
        const settings = getSettings();
        this.openaiClient = null;
        this.geminiClient = null;
        this.groqClient = null;

        // Initialize OpenAI
        if (settings.openaiApiKey) {
            try {
                this.openaiClient = new OpenAI({ apiKey: settings.openaiApiKey });
                console.log('✓ OpenAI client initialized');
            } catch (e) {
                console.error(`✗ OpenAI init failed: ${e.message}`);
            }
        }

        // Initialize Gemini
        if (settings.googleApiKey) {
            try {
                this.geminiClient = new GoogleGenerativeAI(settings.googleApiKey);
                console.log('✓ Gemini client initialized');
            } catch (e) {
                console.error(`✗ Gemini init failed: ${e.message}`);
            }
        }

        // Initialize Groq
        if (settings.groqApiKey) {
            try {
                this.groqClient = new Groq({ apiKey: settings.groqApiKey });
                console.log('✓ Groq client initialized');
            } catch (e) {
                console.error(`✗ Groq init failed: ${e.message}`);
            }
        }
    }

    _getPlatformContext(platform) {
        const platforms = {
            instagram: {
                charLimit: 2200,
                hashtagLimit: 30,
                bestPractices: 'Use emojis, line breaks, and storytelling. End with a CTA.'
            },
            linkedin: {
                charLimit: 3000,
                hashtagLimit: 5,
                bestPractices: 'Professional tone, use bullet points, share insights and value.'
            },
            twitter: {
                charLimit: 280,
                hashtagLimit: 3,
                bestPractices: 'Be concise, use hooks, create threads for longer content.'
            },
            facebook: {
                charLimit: 63206,
                hashtagLimit: 10,
                bestPractices: 'Tell stories, ask questions, encourage engagement.'
            },
            tiktok: {
                charLimit: 2200,
                hashtagLimit: 10,
                bestPractices: 'Trendy, casual, use popular hashtags and hooks.'
            },
            youtube: {
                charLimit: 5000,
                hashtagLimit: 15,
                bestPractices: 'SEO-optimized descriptions, include timestamps and links.'
            }
        };
        return platforms[platform] || platforms.instagram;
    }

    async _generateWithOpenAI(systemPrompt, userPrompt) {
        const response = await this.openaiClient.chat.completions.create({
            model: 'gpt-4o-mini',
            messages: [
                { role: 'system', content: systemPrompt },
                { role: 'user', content: userPrompt }
            ],
            temperature: 0.8,
            max_tokens: 1000,
            response_format: { type: 'json_object' }
        });

        return JSON.parse(response.choices[0].message.content);
    }

    async _generateWithGemini(systemPrompt, userPrompt) {
        const fullPrompt = `${systemPrompt}\n\n${userPrompt}`;
        const model = this.geminiClient.getGenerativeModel({
            model: 'gemini-1.5-flash',
            generationConfig: {
                temperature: 0.8,
                maxOutputTokens: 1000,
                responseMimeType: 'application/json'
            }
        });

        const result = await model.generateContent(fullPrompt);
        const text = result.response.text().trim();
        return this._cleanJsonResponse(text);
    }

    async _generateWithGroq(systemPrompt, userPrompt) {
        const fullSystem = systemPrompt + '\n\nIMPORTANT: Respond ONLY with valid JSON. No markdown, no code blocks, no extra text.';

        const response = await this.groqClient.chat.completions.create({
            model: 'llama-3.1-8b-instant',
            messages: [
                { role: 'system', content: fullSystem },
                { role: 'user', content: userPrompt }
            ],
            temperature: 0.8,
            max_tokens: 1024,
            response_format: { type: 'json_object' }
        });

        const text = response.choices[0].message.content.trim();
        return this._cleanJsonResponse(text);
    }

    _cleanJsonResponse(text) {
        // Remove markdown code blocks if present
        let cleaned = text;
        if (cleaned.startsWith('```json')) {
            cleaned = cleaned.substring(7);
        }
        if (cleaned.startsWith('```')) {
            cleaned = cleaned.substring(3);
        }
        if (cleaned.endsWith('```')) {
            cleaned = cleaned.substring(0, cleaned.length - 3);
        }
        return JSON.parse(cleaned.trim());
    }

    async _generateContent(systemPrompt, userPrompt) {
        // Try OpenAI first
        if (this.openaiClient) {
            try {
                console.log('🤖 Trying OpenAI...');
                return await this._generateWithOpenAI(systemPrompt, userPrompt);
            } catch (e) {
                console.error(`OpenAI failed: ${e.message}`);
            }
        }

        // Fallback to Gemini
        if (this.geminiClient) {
            for (let attempt = 0; attempt < 2; attempt++) {
                try {
                    console.log(`🤖 Using Gemini (attempt ${attempt + 1}/2)...`);
                    if (attempt > 0) {
                        await new Promise(resolve => setTimeout(resolve, 2000));
                    }
                    return await this._generateWithGemini(systemPrompt, userPrompt);
                } catch (e) {
                    console.error(`Gemini attempt ${attempt + 1} failed: ${e.message}`);
                }
            }
        }

        // Final fallback to Groq
        if (this.groqClient) {
            for (let attempt = 0; attempt < 2; attempt++) {
                try {
                    console.log(`🤖 Using Groq (attempt ${attempt + 1}/2)...`);
                    if (attempt > 0) {
                        await new Promise(resolve => setTimeout(resolve, 2000));
                    }
                    return await this._generateWithGroq(systemPrompt, userPrompt);
                } catch (e) {
                    console.error(`Groq attempt ${attempt + 1} failed: ${e.message}`);
                    if (attempt === 1) {
                        throw new Error(`All AI providers failed. Last error: ${e.message}`);
                    }
                }
            }
        }

        throw new Error('No AI provider available. Configure OpenAI, Google, or Groq API key.');
    }

    async generateTextPost(request) {
        const platformCtx = this._getPlatformContext(request.platform);

        const systemPrompt = `You are an expert social media content creator.
Create engaging content for ${request.platform}.

Platform Guidelines:
- Character limit: ${platformCtx.charLimit}
- Hashtag limit: ${platformCtx.hashtagLimit}
- Best practices: ${platformCtx.bestPractices}

Content Requirements:
- Tone: ${request.tone}
- Goal: ${request.goal}
- Include emojis: ${request.includeEmojis}
- Generate hashtags: ${request.includeHashtags}
- Include CTA: ${request.includeCTA}

Respond with valid JSON:
{
    "caption": "The main caption text",
    "hashtags": ["#hashtag1", "#hashtag2"],
    "cta": "Call to action text"
}`;

        const userPrompt = `Create a ${request.tone} social media post for ${request.platform} about:

${request.topic}

Goal: ${request.goal}. Make it engaging and optimized for the platform.`;

        const result = await this._generateContent(systemPrompt, userPrompt);

        const caption = result.caption || '';
        const words = caption.split(' ').length;

        return {
            success: true,
            caption,
            hashtags: request.includeHashtags ? (result.hashtags || []) : [],
            cta: request.includeCTA ? (result.cta || '') : '',
            stats: {
                characters: caption.length,
                words,
                readTime: `${Math.max(1, Math.floor(words / 3))} sec`,
                engagementScore: (result.hashtags?.length >= 5) ? 'High' : 'Medium'
            },
            creditsUsed: 1,
            creditsRemaining: 149
        };
    }

    async generateImagePost(request) {
        const systemPrompt = `You are an expert social media content creator.
Create engaging content for ${request.platform} with an image.

Respond with valid JSON:
{
    "caption": "The caption text",
    "hashtags": ["#hashtag1", "#hashtag2"],
    "cta": "Call to action",
    "imagePrompt": "Detailed prompt for AI image generation"
}`;

        const userPrompt = `Create a ${request.tone} image post for ${request.platform} about:

${request.topic}

Goal: ${request.goal}. Include a detailed image generation prompt.`;

        const result = await this._generateContent(systemPrompt, userPrompt);

        const caption = result.caption || '';
        const words = caption.split(' ').length;

        return {
            success: true,
            caption,
            hashtags: result.hashtags || [],
            cta: result.cta || '',
            imagePrompt: result.imagePrompt || '',
            stats: {
                characters: caption.length,
                words,
                readTime: `${Math.max(1, Math.floor(words / 3))} sec`,
                engagementScore: 'High'
            },
            creditsUsed: 2,
            creditsRemaining: 23
        };
    }

    async generateVideoScript(request) {
        const systemPrompt = `You are an expert video content creator.
Create a ${request.duration} video script for ${request.platform}.

Respond with valid JSON:
{
    "hook": "Attention-grabbing opening line",
    "script": "Main video script content",
    "cta": "Call to action at the end",
    "hashtags": ["#hashtag1", "#hashtag2"]
}`;

        const userPrompt = `Create a ${request.tone} video script for ${request.platform} about:

${request.topic}

Duration: ${request.duration}
Goal: ${request.goal}`;

        const result = await this._generateContent(systemPrompt, userPrompt);

        const script = result.script || '';
        const words = script.split(' ').length;

        return {
            success: true,
            hook: result.hook || '',
            script,
            cta: result.cta || '',
            hashtags: result.hashtags || [],
            stats: {
                characters: script.length,
                words,
                readTime: `${Math.max(1, Math.floor(words / 3))} sec`,
                engagementScore: 'High'
            },
            creditsUsed: 3,
            creditsRemaining: 7
        };
    }
}

// Singleton instance
let openaiService = null;

export function getOpenAIService() {
    if (!openaiService) {
        openaiService = new OpenAIService();
    }
    return openaiService;
}

export default OpenAIService;
